user/languages/cpp.md

```diff
@@ -93,3 +93,134 @@ before_install:
   - test -n $CC  && unset CC
   - test -n $CXX && unset CXX
 ```
+
+### Modern C++ and Compiler Versioning
+
+If you need a more modern toolchain or more granular control over your tools, you will have to handle that yourself. Fortunately, the community has already whitelisted the [repositories](https://github.com/travis-ci/apt-source-whitelist/blob/master/ubuntu.json) and [packages](https://github.com/travis-ci/apt-package-whitelist/blob/master/ubuntu-precise) necessary for such upgrades.
+
+#### GCC on Linux
+
+Ubuntu 12.04 shipped with GCC 4.6.3 and Ubuntu 14.04 shipped with GCC 4.8.2. If you need a more recent version (if you are working with C++11, you *must* do this; support for C++11 was not finished until GCC 4.9), you will want to follow these directions:
+
+```yaml
+matrix:
+  include:
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-4.9
+      env:
+         - MATRIX_EVAL="CXX=g++-4.9"
```

Comment:
The g++-N package depends on gcc-N, e.g. `apt show g++-4.9`:

<pre>
Package: g++-4.9
Version: 4.9.4-2ubuntu1~16.04
Priority: optional
Section: devel
Source: gcc-4.9
Maintainer: Ubuntu Core developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian GCC Maintainers <debian-gcc@lists.debian.org>
Installed-Size: 33.6 MB
Provides: c++-compiler, c++abi2-dev
Depends: gcc-4.9-base (= 4.9.4-2ubuntu1~16.04), gcc-4.9 (= 4.9.4-2ubuntu1~16.04), libstdc++-4.9-dev (= 4.9.4-2ubuntu1~16.04), libc6 (>= 2.14), libcloog-isl4 (>= 0.17), libgmp10 (>= 2:5.0.1~), libisl15 (>= 0.15), libmpc3, libmpfr4 (>= 3.1.3), zlib1g (>= 1:1.1.4)
Suggests: g++-4.9-multilib, gcc-4.9-doc (>= 4.9), libstdc++6-4.9-dbg (>= 4.9.4-2ubuntu1~16.04)
Download-Size: 21.3 MB
APT-Manual-Installed: yes
APT-Sources: http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu xenial/main amd64 Packages
Description: GNU C++ compiler
 This is the GNU C++ compiler, a fairly portable optimizing compiler for C++.
</pre>